File: backend/app/services/protocols/ark/video.py
"""
ArkVideoProtocol — 火山方舟视频生成协议（暂未实现）。

TODO: POST /api/v3/contents/generations 提交，GET /api/v3/contents/generations/tasks/{id} 轮询。
"""

from __future__ import annotations

from app.schemas.result import GenerationResult, PollResult
from app.services.protocols.ark.base import ArkBaseProtocol


class ArkVideoProtocol(ArkBaseProtocol):
    """Ark 视频生成协议（占位）。"""

    protocol_name: str = "ark"

    def build_request(
        self,
        base_url: str,
        api_key: str,
        body: dict,
        capability: str = "video",
    ) -> tuple[str, dict, dict]:
        raise NotImplementedError("Ark 视频生成协议暂未实现")

    def extract_result(self, data: dict, capability: str = "video") -> GenerationResult | None:
        raise NotImplementedError("Ark 视频生成协议暂未实现")

    def supports(self, capability: str) -> bool:
        return capability == "video"

    def extract_task_id(self, data: dict) -> str | None:
        raise NotImplementedError("Ark 视频生成协议暂未实现")

    def build_poll_url(self, base_url: str, upstream_task_id: str) -> str:
        raise NotImplementedError("Ark 视频生成协议暂未实现")

    def parse_poll_response(self, data: dict, capability: str = "video") -> PollResult:
        raise NotImplementedError("Ark 视频生成协议暂未实现")

File: server/services/protocols/openai/video.ts
/**
 * OpenAI 视频生成协议。
 * 构建视频生成的上游请求；响应/轮询解析逻辑在 shared.ts（与 image 共用）。
 */

import type {
  ProtocolRequestResult,
  ProtocolResponse,
  ProtocolService,
  PollResult,
} from "@server/services/protocols/base";
import {
  parseScanPollResult,
  parseScanSyncResult,
  extractOpenAiTaskId,
  buildOpenAiPollUrl,
} from "./shared";

/** 裸 base64 产物补通用前缀（MIME 省略，由播放器按内容识别） */
const B64_MIME = "data:;base64,";

export class OpenAiVideoProtocol implements ProtocolService {
  readonly name = "openai_video";

  buildVideoRequest(
    baseUrl: string,
    apiKey: string,
    body: Record<string, unknown>,
    channelConfig?: Record<string, unknown>
  ): ProtocolRequestResult {
    const endpoints = (channelConfig?.protocol as Record<string, unknown>)?.endpoints as Record<string, string> | undefined;
    const endpoint = endpoints?.["video.generations"] ?? "/videos";

    return {
      url: `${baseUrl}${endpoint}`,
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body,
    };
  }

  parseVideoResponse(response: unknown): ProtocolResponse {
    return parseScanSyncResult(response, B64_MIME);
  }

  // 异步任务支持

  extractTaskId(data: unknown, channelConfig?: Record<string, unknown>, capability?: string): string | null {
    return extractOpenAiTaskId(data, channelConfig, capability ?? "video");
  }

  buildPollUrl(baseUrl: string, upstreamTaskId: string, channelConfig?: Record<string, unknown>, capability?: string, model?: string): string {
    return buildOpenAiPollUrl(baseUrl, upstreamTaskId, channelConfig, capability ?? "video", model);
  }

  parsePollResponse(data: unknown): PollResult {
    return parseScanPollResult(data, B64_MIME);
  }
}
